src/lib/uri/uri.h

```diff
@@ -154,6 +154,14 @@ uri_escape(const char *src, size_t src_size, char *dst,
 size_t
 uri_unescape(const char *src, size_t src_size, char *dst, bool decode_plus);
 
+/** Determine if uris refer to the same host:service or unix socket path. */
+bool
+uri_equals(const struct uri *a, const struct uri *b);
+
+/** Check if a uri is empty. */
+bool
+uri_is_nil(const struct uri *uri);
```

Comment:
You might need new tests in `test/unit/uri.c`.

Reply:
Sure:
<details><summary>diff</summary>

```diff
diff --git a/src/lib/uri/uri.c b/src/lib/uri/uri.c
index 4e0792bc5..07551cc53 100644
--- a/src/lib/uri/uri.c
+++ b/src/lib/uri/uri.c
@@ -483,5 +483,10 @@ uri_equals(const struct uri *a, const struct uri *b)
 bool
 uri_is_nil(const struct uri *uri)
 {
+	/*
+	 * Check only these 3 fields, because without them a uri doesn't make
+	 * sense. But technically this will give some false positives. For
+	 * example, for uris with non-empty fragment or query.
+	 */
 	return uri->host == NULL && uri->path == NULL && uri->service == NULL;
 }
diff --git a/test/unit/uri.c b/test/unit/uri.c
index 34961d6c6..97386d85f 100644
--- a/test/unit/uri.c
+++ b/test/unit/uri.c
@@ -54,7 +54,7 @@ empty_uri_create(struct uri *uri)
 static void
 empty_uri_check(struct uri *uri, const char *msg)
 {
-	plan(11);
+	plan(12);
 	is(uri->scheme, NULL, "%s scheme", msg);
 	is(uri->login, NULL, "%s login", msg);
 	is(uri->password, NULL, "%s password", msg);
@@ -66,6 +66,7 @@ empty_uri_check(struct uri *uri, const char *msg)
 	is(uri->host_hint, 0, "%s hint", msg);
 	is(uri->param_count, 0, "%s param count", msg);
 	is(uri->params, NULL, "%s params", msg);
+	ok(uri_is_nil(uri), "%s is_nil()", msg);
 	check_plan();
 }
 
@@ -137,6 +138,53 @@ test_move_empty(void)
 	footer();
 }
 
+struct uri_equal_expected {
+	/** Source string for the first uri. */
+	const char *src_a;
+	/** Source string for the second uri. */
+	const char *src_b;
+	/** Expected comparison result. */
+	bool is_equal;
+};
+
+static void
+test_equals(void)
+{
+	struct uri_equal_expected test_pairs[] = {
+		{NULL, NULL, true},
+		{"localhost", "localhost", true},
+		{"user@localhost", "localhost", true},
+		{"user:pass@localhost", "localhost", true},
+		{"user:pass@localhost", "user@localhost", true},
+		{"localhost:3301", "localhost:3302", false},
+		{"host_a", "host_b", false},
+		{"scheme://localhost", "localhost", true},
+		{"scheme1://host:port", "scheme2://host:port", true},
+		{"localhost/path/to/file", "localhost", false},
+		{"/path/to/file", "/path/to/file", true},
+		{"/path/to/file", "localhost/path/to/file", false},
+		{"unix/path/to/file", "/path/to/file", false},
+		{"unix/:/path/to/file", "/path/to/file", true},
+	};
+	header();
+	plan(3 * lengthof(test_pairs));
+	for (unsigned i = 0; i < lengthof(test_pairs); i++) {
+		struct uri uri_a, uri_b;
+		const char *src_a = test_pairs[i].src_a;
+		const char *src_b = test_pairs[i].src_b;
+		bool is_equal = test_pairs[i].is_equal;
+		ok(uri_create(&uri_a, src_a) == 0, "uri_create(%s)",
+		   src_a ? src_a : "NULL");
+		ok(uri_create(&uri_b, src_b) == 0, "uri_create(%s)",
+		   src_b ? src_b : "NULL");
+		is(uri_equals(&uri_a, &uri_b), is_equal, "%s %s equal to %s",
+		   src_a ? src_a : "NULL", is_equal ? "is" : "isn't",
+		   src_b ? src_b : "NULL");
+	}
+	check_plan();
+	footer();
+}
+
 struct uri_param_expected {
 	/** URI parameter name */
 	const char *name;
@@ -780,11 +828,12 @@ test_unescape_special_cases(void)
 int
 main(void)
 {
-	plan(10);
+	plan(11);
 	test_copy_sample();
 	test_copy_empty();
 	test_move_sample();
 	test_move_empty();
+	test_equals();
 	test_string_uri_with_query_params_parse();
 	test_string_uri_set_with_query_params_parse();
 	test_invalid_string_uri_set();
```
</details>